Scripts to build on windows

josecv opened the pull request:
## Description

A few sentences describing the overall goals of the pull request's commits.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [x] My change is adequately tested.
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

josecv committed:
bump ko

Signed-off-by: Jose Cortes <josecortes@datawire.io>

thallgren reviewed (approved):
Assuming it all works, this looks good to me. I have a hard time parsing those for loops ;-)

thallgren commented:
Oh, and you'll need to gpg-sign and sign-off on the commits.

josecv committed:
Add a script to create kubeception test clusters

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Wrap errors in rotatingfile.go

Especially important to do this because any errors there happen before
logging is setup, so we need to know where they're coming from.

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Prevent windows.ERROR_SUCCESS from being returned from SetOwnerAndGroup

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Use WMI to set the TUN device's DNS domain

This works around an issue with the tun device's SetDNS method where on
some systems hosts under the device's domain won't be resolved (e.g.
resolving hello-0 to hello-0.default.svc.cluster.local)

Signed-off-by: Jose Cortes <j14a@protonmail.com>

Commit:
Refactor install tests into a test suite

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Some enhancements to the intercept suite's response check test.

This helps disambiguate different classes of test failures

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Require (don't assert) that list succeeds before doing os.Stat()

And also add a delay before the stat call.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that mount points are concurrency protected

In situations where an intercepted pod dies followed by another
pod taking over, it is important that the file system is completely
unmounted before an attempt is made to mount it again.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv committed:
Fix tests to run on windows

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Build a docker image for tests, load it into kubeception kluster

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Add a script to wait until a kubernetes cluster is up

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Small fix to allow telepresence-test-developer to see pod logs

This should ensure that we always get the traffic manager's logs in
CircleCI

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Log every telepresence command the tests run

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Bump TELEPRESENCE_MAX_LOGFILES to ensure we always have daemon logs.

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Add windows tests to CircleCI configuration

Signed-off-by: Jose Cortes <josecortes@datawire.io>

Commit:
Improved error handling when removing mountpoints

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv committed:
A script to build windows stuff via docker

Signed-off-by: Jose Cortes <josecortes@datawire.io>

Commit:
Clean up chatty logging

Moves some logging from INFO to DEBUG level and removes logging of a
couple of errors that aren't really errors.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung reviewed (approved):
This LGTM and since I had a PR that was built on top of this PR, I anecdotally feel that the tests are very thorough and consistent.

donnyyung commented on .circleci/config.yml:
```diff
@@ -108,6 +176,51 @@ commands:
+          platform: windows
+      - save-logs
+
   "build-and-test":
```
I would maybe change this to "build-and-test-linux" but we can always do that at a later point

LukeShu commented on pkg/client/logging/dup_unix.go:
```diff
@@ -21,3 +23,23 @@ func dupToStd(file *os.File) (err error) {
 	return err
 }
+
+func dupStd() (int, int, error) {
+	stdoutFd, err := unix.Dup(1)
+	if err != nil {
+		return -1, -1, err
+	}
+	stderrFd, err := unix.Dup(2)
+	if err != nil {
+		return -1, -1, err
+	}
+	return stdoutFd, stderrFd, err
+}
+
+func restoreStd(stdoutFd, stderrFd int) error {
+	err := unix.Dup2(stdoutFd, 1)
+	if err != nil {
+		return err
+	}
+	return unix.Dup2(stderrFd, 2)
+}
```
Since these are only used by the tests, I'd prefer for them to be in `dup_{unix,windows}_test.go` files, rather than in the main `dup_unix.go` and `sysinfo_windows.go`.

LukeShu commented on pkg/client/logging/sysinfo_windows.go:
```diff
@@ -16,6 +16,14 @@ func dupToStd(_ *os.File) interface{} {
 	return errors.New("dupToStd() is not implemented on windows")
 }
 
+func dupStd() (int, int, error) {
+	return -1, -1, errors.New("dupStd() is not implemented on windows")
+}
+
+func restoreStd(int, int) error {
+	return errors.New("restoreStd() is not implemented on windows")
+}
+
```
Rather than having `initcontext_test.go` wrap these calls in `if runtime.GOOS != "windows" { … }`, why not just make these be safe no-ops on Windows?  And the instead of having them return ints, return a cleanup function.

So on unix something like

```go
func dupStd() (func(), error) {
	stdoutFd, err := unix.Dup(1)
	if err != nil {
		return nil, err
	}
	stderrFd, err := unix.Dup(2)
	if err != nil {
		return nil, err
	}
	return func() {
		_ = unix.Dup2(stdoutFd, 1)
		_ = unix.Dup2(stderrFd, 2)
	}, nil
}
```

and on Windows

```go
func dupStd() (func(), error) {
	return func() {}, nil
}
```

josecv replied:
that is definitely much better

LukeShu commented on pkg/client/logging/rotatingfile.go:
```diff
@@ -240,48 +240,48 @@ func (rf *RotatingFile) openNew(prevInfo SysInfo) (err error) {
 	if rf.file == nil {
 		if newFile, err = createFile(fullPath, rf.fileMode); err != nil {
-			return err
+			return fmt.Errorf("Failed to createFile %s: %w", fullPath, err)
```
Error messages usually start lowercase

LukeShu commented on winmake.bat:
```diff
@@ -0,0 +1,46 @@
+echo off
```
I'd encourage you to put files like this in the `build-aux/` directory.

LukeShu commented on Dockerfile.winbuild:
```diff
@@ -0,0 +1,17 @@
+FROM golang:1.16.5
```
I'd encourage you to put files like this in the `build-aux/` directory.

LukeShu commented on pkg/client/cli/telepresence_test.go:
```diff
@@ -171,11 +173,13 @@ func (ts *telepresenceSuite) TearDownSuite() {
+			ts.Empty(stderr)
+			ts.Contains(stdout, fmt.Sprintf("Client: %s", client.DisplayVersion()))
+		})
+	}
```
Why can't this test run on Windows?

josecv replied:
it actually can now, we just had to update the stable.txt file; good catch, i'll re-enable this test

LukeShu replied:
And even then, it would be better to call `.SkipNow()` inside of the test.

LukeShu commented on pkg/client/cli/telepresence_test.go:
```diff
@@ -303,6 +307,11 @@ func (ts *telepresenceSuite) TestA_WithNoDaemonRunning() {
+		if goRuntime.GOOS == "windows" {
+			// Windows needs some time to get its bearings
+			time.Sleep(10 * time.Second)
+		}
```
This at least needs a "FIXME" or a "BUG" comment; `telepresence connect` should block until it is set up.

LukeShu commented on pkg/client/cli/telepresence_test.go:
```diff
@@ -619,6 +637,10 @@ func (cs *connectedSuite) TestJ_ListOnlyMapped() {
+	// This test only runs on linux as it requires docker, and CI can't run linux docker containers inside non-linux runners
+	if goRuntime.GOOS != "linux" {
+		return
+	}
```
Prefer to use `cs.SkipNow()` (`cs.T().SkipNow()`?) instead of just returning early, so that the summary shows that it was skipped.

Does CI have `docker` installed?  If so, we should prefer to make this decision based on that rather than coding CI details in to the tests.

josecv replied:
So CI has `docker` installed, it's just that pulling the image will fail because there's no windows version available.

I'll make the skipNow change

LukeShu commented on pkg/client/cli/telepresence_test.go:
```diff
@@ -1391,17 +1441,22 @@ func (ts *telepresenceSuite) kubectlOut(ctx context.Context, args ...string) (st
-	includeEnv := []string{"HOME=", "PATH=", "LOGNAME=", "TMPDIR=", "MAKELEVEL="}
+	includeEnv := []string{"HOME=", "PATH=", "Path=", "LOGNAME=", "TMPDIR=", "MAKELEVEL="}
 	for _, env := range os.Environ() {
+		dlog.Infof(ctx, "I found variable %s", env)
```
Not OK! This will leak CI secrets if the test fails!

josecv replied:
OMG this is very stupid on my part; I believe Circle should [mask](https://circleci.com/docs/2.0/env-vars/#secrets-masking) the secrets, and I haven't seen them be printed, but I'll definitely remove this and go check that it hasn't  been printed anywhere

LukeShu commented on pkg/client/logging/initcontext_test.go:
```diff
@@ -98,24 +99,27 @@ func TestInitContext(t *testing.T) {
+	// This will fail on Windows
+	if runtime.GOOS != "windows" {
+		t.Run("captures output of builtin functions", func(t *testing.T) {
+			ctx, _, logFile := testSetup(t)
```
It would be better to `.SkipNow()` inside of the test.

LukeShu commented on .circleci/config.yml:
```diff
@@ -7,9 +7,13 @@
 #     + DOCKERHUB_PASSWORD
 #     + AWS_ACCESS_KEY_ID (for S3)
 #     + AWS_SECRET_ACCESS_KEY (for S3)
+#     + TELEPRESENCE_KUBECEPTION_TOKEN
```
Variables that Telepresence isn't using at runtime should probably have `DEV_` in them.

LukeShu commented on pkg/client/cli/telepresence_test.go:
```diff
@@ -1526,6 +1584,9 @@ func telepresence(t testing.TB, args ...string) (string, string) {
 	var stdout, stderr strings.Builder
+	// The cmd object does not log with timestamps, so dump out the telepresence command here.
+	// That way we have a timestamped record of when it ran, which is useful for correlating with the daemon logs
+	dlog.Debug(ctx, "telepresence invoked with", args)
```
Bwuuu? `cmd` is using dlog, same as this is?

LukeShu commented on .circleci/config.yml:
```diff
@@ -108,22 +176,74 @@ commands:
+      - attach_workspace:
+          at: /tmp/workspace
+      - prepare-kluster:
+          platform: linux
```
Part of me doesn't like that we're doing this even when unnecessary (i.e. on GOOS=linux where dtest can spin up a local k3s cluster just fine).  But another part of me hates dtest and is excited that this gets us closer to killing it.

LukeShu commented on pkg/tun/device_windows.go:
```diff
@@ -102,7 +112,40 @@ func (t *Device) setDNS(ctx context.Context, server net.IP, domains []string) (e
+	if err := cmd.Run(); err != nil {
+		// Log the error, but don't actually fail on it: This is all just a fallback for SetDNS, so the domains might actually be working
+		dlog.Errorf(ctx, "Failed to set NetworkAdapterConfiguration DNS Domain: %v. Will proceed, but namespace mapping might not be functional.", err)
+	}
```
gross gross gross :)

I'm not looking at this too closely, but it would be good to include references in those comments ("apparently well-known"), and maybe be more specific that "some systems".

josecv replied:
> would be good to include references in those comments ("apparently well-known"),

Good point, I'll add some links to the comment

> and maybe be more specific that "some systems".

This I can't do :( unfortunately "some systems" means "CircleCI but not Jose's windows computer"

LukeShu reviewed (changes requested):
A couple of minor changes.

josecv committed:
Address PR comments

Signed-off-by: Jose Cortes <josecortes@datawire.io>
